Initial Vita + Switch Merger

halotroop2288 opened the pull request:
This rebases https://github.com/SonicMastr/godot-vita/tree/3.5 onto the latest commit of https://github.com/godotengine/godot/tree/3.5 and merges in https://github.com/Stary2001/godot/tree/3.5.2-rc_switch by combining any conflicting `#ifdef`s into one line where possible.

Every `#endif` and `#else` in the affected files (to my knowledge) has been commented with its starting conditions so that they will conflict properly and prevent future devs from seeing cryptic errors about things that are accidentally disabled.

I also created a higher-quality logo for the Switch exporter, because it didn't look good on dark themes before.

Stary2001 committed:
Initial switch thing

Stary2001 committed:
thirdparty fixes

Stary2001 committed:
Add switch platform files, fix cflags etc..

Stary2001 committed:
Add a whole bunch of driver stuff, it works now. Turns out OS::swap_buffers is pretty important.

Stary2001 committed:
Touch support on switch, updated GL context creation for latest mesa/nouveau release

Stary2001 committed:
Updated EGL stuff for latest mesa/nouveau/libnx etc, fixed get_ticks_usec, fixed threadCreate. Godot hangs.

Stary2001 committed:
Make detect properly detect missing system pkg-config

Stary2001 committed:
switch to pthreads, switch to wrapping libnx instead of horrific sed everywhere. also fixed gl context deinit

Stary2001 committed:
update for newest godot

Stary2001 committed:
remove egl debug

Stary2001 committed:
icons are fine actually. disable safe save.

Cpasjuste committed:
switch: use posix (pthread) locks and semaphores

Cpasjuste committed:
switch: remove _POSIX_READER_WRITER_LOCKS defines, as origin

Cpasjuste committed:
switch: start working on export

Cpasjuste committed:
switch: .pck exports working

Cpasjuste committed:
switch: minor cleanup

Cpasjuste committed:
switch: add joypads support

Cpasjuste committed:
switch: add joypads support, oups

Stary2001 committed:
oops

Stary2001 committed:
fix sysv_abi warning spam, detect.py update

Cpasjuste committed:
switch: minor cleanup

Cpasjuste committed:
switch: add joypads axis support

Cpasjuste committed:
switch: fix get_executable_path, this allow pck loading

Cpasjuste committed:
switch: add audio driver

Cpasjuste committed:
switch: fix code indentation

Cpasjuste committed:
switch: register 8 joypads by default
switch: go.sh: build release target, add template export

Cpasjuste committed:
switch: enable etc2 texture compression feature in export template
switch: "go.sh" to build debug target, add "go_release.sh" to build release targets

Cpasjuste committed:
switch: some cleanup in export template

Stary2001 committed:
Update to enable webm module.

Cpasjuste committed:
switch: update icon, title, author and version on export. WIP

Cpasjuste committed:
switch: finish icon/nacp export

Cpasjuste committed:
switch: check icon size on export

fhidalgosola committed:
cleanup of build scripts

Stary2001 committed:
switch: enable upnp

Stary2001 committed:
Readme!

Stary2001 committed:
Change mirror URL.

Stary2001 committed:
fix python, disable nso for now

Stary2001 committed:
Add applet splash

Stary2001 committed:
Disable s3tc support for now.

Stary2001 committed:
fix platform_config.h being lost in rebase

Stary2001 committed:
Brings the Switch platform up to 3.2 compat.

Stary2001 committed:
oops

Stary2001 committed:
try to support multitouch/fix weirdness with key_touch based impl

Stary2001 committed:
fix python2/3 compat issue

Stary2001 committed:
Add CI.

Stary2001 committed:
with the switch to wslay, we can reenable websockets. also enable system libzstd

Stary2001 committed:
make delay_usec actually delay properly

Stary2001 committed:
add nxlink one-click export

Stary2001 committed:
Initial keyboard support.

Stary2001 committed:
whitespace

Stary2001 committed:
...

MikeTheSnowman committed:
Updated get_virtual_keyboard_height 

Updated get_virtual_keyboard_height to return 0 if the virtual keyboard is not opened.

Stary2001 committed:
redo export a bit

Stary2001 committed:
update before doing linux ci, update readme

Stary2001 committed:
don't upgrade, because it pulls in every recent version of php or something

Stary2001 committed:
redo nro updating code

Stary2001 committed:
export: actually set romfs offset, switch: always romfsinit

Stary2001 committed:
load pck out of romfs where available

Stary2001 committed:
use new app_args fileassoc feature to make it actually work

Stary2001 committed:
disable fused builds by default, add custom editor id support + override export_pack

Stary2001 committed:
testing ci in production, what could go wrong?

Stary2001 committed:
fix ci ids

Stary2001 committed:
actions machine broke

Stary2001 committed:
use release target

Stary2001 committed:
Revert "use release target"

This reverts commit b235052c9e3bc71523f17167f9e2e84c23830467.

Stary2001 committed:
template build when this ita gged, hopefully

Stary2001 committed:
fix hbmenu cfg, fix actions hopefully

Stary2001 committed:
hopefully fix godot nro filename

Stary2001 committed:
fix double free

Stary2001 committed:
fix audio driver for switch

Stary2001 committed:
whoops dont build appimage for now

Stary2001 committed:
no, really

Stary2001 committed:
dont syu, add version.txt to template zip

Stary2001 committed:
let's give this a go then

Stary2001 committed:
fug

Stary2001 committed:
adjust workflow, maybe it works

Stary2001 committed:
fix it

Stary2001 committed:
add nxlink warning, try bundled binaries first

Stary2001 committed:
Modernise CI config. This also runs formatters!

Stary2001 committed:
Whoops, *.exe -> .exe. Fixes Windows executable bundling.

Stary2001 committed:
fix readme...

Stary2001 committed:
update for 3.3 code changes

Stary2001 committed:
run clang-format

Stary2001 committed:
make export more sound...

EIREXE committed:
Fix user directory access

EIREXE committed:
Force never using RENDER_SEPARATE_THREAD because it's broken on switch

Stary2001 committed:
add new plist

Stary2001 committed:
switch: prepend devkitA64 to path rather than append, fixes build with aarch64-none-elf gcc already present

Stary2001 committed:
switch: update to new hid api

Stary2001 committed:
switch: build nacp as needed, put the engine version in it (this will be overriden by the editor for fused builds)

Stary2001 committed:
fix nacp target...

Stary2001 committed:
appease code style

SonicMastr committed:
3.5 Port (Input not working/everything else is)

SonicMastr committed:
Vita: Fixed Input (There was nothing wrong)

SonicMastr committed:
Starting on exporter

SonicMastr committed:
Vita: Finish Exporter

SonicMastr committed:
Vita: try to pass static checks

SonicMastr committed:
Vita: Fix exporter building

SonicMastr committed:
Thanks Clang

SonicMastr committed:
Vita: Test Linux Error

SonicMastr committed:
Vita: Test Linux Error

SonicMastr committed:
clang_format

SonicMastr committed:
Merge branch '3.x' of https://github.com/SonicMastr/godot into 3.x

SonicMastr committed:
Vita: rc5 for the sake of templates

SonicMastr committed:
Vita: Fix up the exporter a bit

SonicMastr committed:
Merge branch 'godotengine:3.x' into 3.x

SonicMastr committed:
Vita: Add Accelerometer/Gyroscope Support

SonicMastr committed:
Vita: Fix MSAA

SonicMastr committed:
Vita: Implement WebSocket

SonicMastr committed:
Fix Windows Build

SonicMastr committed:
Merge branch 'godotengine:3.x' into 3.x

SonicMastr committed:
Vita: Change default memory split

SonicMastr committed:
Merge remote-tracking branch 'upstream/3.x' into 3.x

SonicMastr committed:
Vita: Fix compressed texture NPOT

SonicMastr committed:
Vita: Add PVRTC (Experimental)

SonicMastr committed:
Vita: Add OSK support

SonicMastr committed:
Vita: Add OSK support (cleanup)

SonicMastr committed:
Merge branch 'godotengine:3.x' into 3.x

SonicMastr committed:
Vita: Adjust build checks

SonicMastr committed:
Vita: Add actions (works locally)

SonicMastr committed:
Cleanup

SonicMastr committed:
Revert hardcoded status

SonicMastr committed:
3.5 Update to Upstream

3.5 Update to Upstream

SonicMastr committed:
Vita: Update Base Branch

SonicMastr committed:
Merge branch 'godotengine:3.5' into 3.5

SonicMastr committed:
Merge remote-tracking branch 'upstream/3.5' into 3.5

Stary2001 committed:
code style

Stary2001 committed:
Update Switch platform for 3.5 support

Stary2001 committed:
stub read_link, create_link on switch

Stary2001 committed:
hopefully fix switch ci

SonicMastr committed:
Merge branch 'godotengine:3.5' into 3.5

Stary2001 committed:
disable Mono for Linux editor

Stary2001 committed:
Add temporaries in create_nacp - fix UB. This fixes the weird "title becoming version/name" bug.

SonicMastr committed:
Merge branch 'godotengine:3.5' into 3.5

Stary2001 committed:
Flush input events. This is the only thing that broke the new version. Really.

Stary2001 committed:
Update copyright headers with clang-format.

Stary2001 committed:
Wait, how did this ever work?

Stary2001 committed:
Don't crash on empty title/author/version field

SonicMastr committed:
Merge branch 'godotengine:3.5' into 3.5

halotroop2288 committed:
Merge remote-tracking branch 'Stary2001(Switch)/3.5.2-rc_switch' into 3.5-vita-switch

# Conflicts:
#	drivers/unix/dir_access_unix.cpp
#	drivers/unix/file_access_unix.cpp
#	drivers/unix/ip_unix.cpp
#	drivers/unix/ip_unix.h
#	drivers/unix/net_socket_posix.cpp

halotroop2288 committed:
Label your `#else`s and `#endif`s!

The amount of time that I have spent fixing issues caused by uncommented #`endif`s being merged as if there's no conflict is absurd.

halotroop2288 committed:
New Switch logo

The old one with a solid white background looked pretty bad on a dark-themed editor.

halotroop2288 committed:
Format Python scripts

halotroop2288 committed:
CLANG!

halotroop2288 commented:
CI Success! Now I might change the scripts a bit so they don't all go off during every platform-specific change.

![CI Success](https://media.discordapp.net/attachments/640231384971083807/1093801866657275944/image.png?width=826&height=479)

halotroop2288 committed:
[CI] Ignore Changes to Unrelated Paths Mk.1

Stary2001 commented on README.md:
Readme should be rewritten. Probably we need a wiki.

halotroop2288 replied:
True. I had done this before but it got deleted when I reset on this project.

- [x] Rewrite README.md

halotroop2288 replied:
> Probably we need a wiki.

I'll fork the official wiki later so we can have good documentation.

Stary2001 commented on drivers/gles2/rasterizer_storage_gles2.cpp:
```diff
@@ -6239,12 +6246,12 @@ void RasterizerStorageGLES2::initialize() {
 	// S3TC it will crash trying to load these textures, as they are not exported in the APK. This is a simple way
 	// to prevent Android devices trying to load S3TC, by faking lack of hardware support.
-#if defined(ANDROID_ENABLED) || defined(IPHONE_ENABLED)
+#if defined(ANDROID_ENABLED) || defined(IPHONE_ENABLED) || defined(VITA_ENABLED)
```
This is happening in two different places for vita/switch.

halotroop2288 replied:
- [x] Add `|| HORIZON_ENABLED` to this block

Stary2001 commented on drivers/gles2/rasterizer_storage_gles2.cpp:
```diff
@@ -6294,6 +6301,17 @@ void RasterizerStorageGLES2::initialize() {
+	// Switch: this happens on Horizon too.
+#ifndef TOOLS_ENABLED
+#if defined ANDROID_ENABLED || defined HORIZON_ENABLED
+	config.s3tc_supported = false;
```
This is happening in two different places.

halotroop2288 replied:
- [x] Remove this block

Stary2001 commented on drivers/unix/dir_access_unix.cpp:
```diff
@@ -405,14 +418,21 @@ bool DirAccessUnix::is_link(String p_file) {
 }
 
 String DirAccessUnix::read_link(String p_file) {
+#ifdef VITA_ENABLED
```
Maybe reorganise this to have if defined(vita) || if defined(horizon)

halotroop2288 replied:
Would that have the same results? Then OK.

- [x] Add `|| HORIZION_ENABLED` to this block

Stary2001 commented on editor/export_template_manager.cpp:
```diff
@@ -230,7 +230,7 @@ void ExportTemplateManager::_refresh_mirrors() {
 
 	String current_version = VERSION_FULL_CONFIG;
-	const String mirrors_metadata_url = "https://godotengine.org/mirrorlist/" + current_version + ".json";
+	const String mirrors_metadata_url = "https://9net.org/godot/" + current_version + ".json";
```
Careful! I can host the Vita templates if needed.

halotroop2288 replied:
Having one place for this stuff would be great!

Stary2001 commented on modules/bullet/SCsub:
```diff
@@ -200,6 +200,11 @@ if env["builtin_bullet"]:
     thirdparty_sources = [thirdparty_dir + file for file in bullet2_src]
 
     env_bullet.Prepend(CPPPATH=[thirdparty_dir])
+    # Treat Bullet headers as system headers to avoid raising warnings. Not supported on MSVC. (TODO SWITCH: devkitA64 issue prevents this from working)
```
This is mostly a reminder for me, to look at this later.

halotroop2288 replied:
Does this require a change?

Stary2001 commented on editor/editor_settings.cpp:
```diff
@@ -381,7 +381,7 @@ void EditorSettings::_load_defaults(Ref<ConfigFile> p_extra_config) {
 	// On save
 	_initial_set("filesystem/on_save/compress_binary_resources", true);
-	_initial_set("filesystem/on_save/safe_save_on_backup_then_rename", true);
+	_initial_set("filesystem/on_save/safe_save_on_backup_then_rename", false);
```
Pretty sure I did this while trying to get editor working on Switch. Should probably be reverted.

halotroop2288 replied:
- [x] false -> true

Stary2001 commented on .gitignore:
```diff
@@ -86,6 +86,17 @@ platform/android/java/*/libs/
+
+# Switch
+*.nacp
+*.nca
```
Why? We don't generate these.

halotroop2288 replied:
It's useful to have any platform-specific binaries in the list in case they somehow end up there. I didn't expect to see `.nacp` files, but I found some in the repo. So I'm just being thorough.

Stary2001 commented on drivers/gles2/shaders/canvas.glsl:
```diff
@@ -310,20 +310,20 @@ VERTEX_SHADER_CODE
-#extension GL_EXT_shader_texture_lod : enable
-#define texture2DLod(img, coord, lod) texture2DLodEXT(img, coord, lod)
-#define textureCubeLod(img, coord, lod) textureCubeLodEXT(img, coord, lod)
+//#extension GL_EXT_shader_texture_lod : enable
```
I'm concerned about this - are the extensions "supported" but not working on Vita?

SonicMastr replied:
This is exactly the case. There's a better way to handle this but since at the time I was only worried about the vita (and it didn't change actual functionality of the shader on the other platforms), I never made vita specific defines. That shouldn't be too hard to add.

halotroop2288 replied:
Commented code is always very concerning. This is a question for @SonicMastr.

- [x] Uncomment this code into a preprocessor block

halotroop2288 replied:
> There's a better way to handle this

@SonicMastr Besides preprocessors, what were you thinking of?

Stary2001 commented on platform/vita/app/module/libGLESv2.suprx:
Should this be in the repo?

SonicMastr replied:
Yes. This is here for version control and to make sure no matter what I know exactly which version of the built module is in use. (And for template building)

Stary2001 commented on scripts/create-switch-release.sh:
I said before about this being something I used in old CI so I didn't use it locally. At least remove the apt-get update, maybe auto grab the version number from version.py.

halotroop2288 replied:
Auto-grabbing the version in bash is not something I know how to do. Release and template scripts are useful for testing though, not just CI which I would prefer to have releases in anyway.

Stary2001 commented on scripts/create-template.sh:
Same re apt-get, version number

Stary2001 reviewed (commented):
I've given this an initial review. I didn't look too hard at the Vita specific code, but tried to look at the core changes.

halotroop2288 commented on drivers/gles2/rasterizer_storage_gles2.cpp:
```diff
@@ -89,18 +89,20 @@ GLuint RasterizerStorageGLES2::system_fbo = 0;
 //void *glResolveMultisampleFramebufferAPPLE;
 #define glRenderbufferStorageMultisample glRenderbufferStorageMultisampleAPPLE
-#elif defined(ANDROID_ENABLED)
+#elif defined ANDROID_ENABLED || defined HORIZON_ENABLED
```
Uh... I don't think this is correct syntax, is it?
I mean, I did test builds, but I thought there had to be parentheses around these...

Stary2001 replied:
per [cppreference](https://en.cppreference.com/w/cpp/preprocessor/conditional):
> The expression may contain:
> unary operators in form defined identifier or defined (identifier)

So this _is_ valid, but it's best to be consistent.

halotroop2288 commented on drivers/gles2/shaders/cubemap_filter.glsl:
```diff
@@ -28,20 +28,20 @@ void main() {
-#extension GL_EXT_shader_texture_lod : enable
-#define texture2DLod(img, coord, lod) texture2DLodEXT(img, coord, lod)
-#define textureCubeLod(img, coord, lod) textureCubeLodEXT(img, coord, lod)
+//#extension GL_EXT_shader_texture_lod : enable
```
- [x] Uncomment this code into a preprocessor block

halotroop2288 commented on drivers/gles2/shaders/effect_blur.glsl:
```diff
@@ -39,20 +39,20 @@ void main() {
-#extension GL_EXT_shader_texture_lod : enable
-#define texture2DLod(img, coord, lod) texture2DLodEXT(img, coord, lod)
-#define textureCubeLod(img, coord, lod) textureCubeLodEXT(img, coord, lod)
+//#extension GL_EXT_shader_texture_lod : enable
```
- [x] Uncomment this code into a preprocessor block

halotroop2288 commented on drivers/gles2/shaders/scene.glsl:
```diff
@@ -222,19 +222,19 @@ void light_compute(
+		m_var = 0.662002687 * S1 + 0.684122060 * S2 - 0.323583601 * S3 - 0.0225411470 * m_var;\
+		}
+	*/
+	//#define SRGB_APPROX(m_var)
```
- [x] Uncomment this code into a preprocessor block

halotroop2288 commented on drivers/gles2/shaders/scene.glsl:
```diff
@@ -729,20 +729,20 @@ VERTEX_SHADER_CODE
-#extension GL_EXT_shader_texture_lod : enable
-#define texture2DLod(img, coord, lod) texture2DLodEXT(img, coord, lod)
-#define textureCubeLod(img, coord, lod) textureCubeLodEXT(img, coord, lod)
+//#extension GL_EXT_shader_texture_lod : enable
```
- [x] Uncomment this code into a preprocessor block

halotroop2288 commented on drivers/gles2/shaders/tonemap.glsl:
```diff
@@ -30,20 +30,20 @@ void main() {
-#extension GL_EXT_shader_texture_lod : enable
-#define texture2DLod(img, coord, lod) texture2DLodEXT(img, coord, lod)
-#define textureCubeLod(img, coord, lod) textureCubeLodEXT(img, coord, lod)
+//#extension GL_EXT_shader_texture_lod : enable
```
- [x] Uncomment this code into a preprocessor block

halotroop2288 commented on drivers/unix/dir_access_unix.cpp:
```diff
@@ -405,14 +418,21 @@ bool DirAccessUnix::is_link(String p_file) {
 
 	p_file = fix_path(p_file);
 
+#ifdef HORIZON_ENABLED
```
- [x] Remove this block

halotroop2288 committed:
Add Switch-specific intermediary build files to .gitignore

halotroop2288 committed:
Remove apt-get from Switch scripts

halotroop2288 committed:
Consistent `#if defined(FOO) || defined(bar)`

And labeled `#else` and `#endif` in `rasterizer_storage_gles2.cpp`

halotroop2288 committed:
Enable safe save on backup then remove by default

halotroop2288 committed:
Fix preprocessors in GL shaders

- Uncommented code from the Vita branch.
- Labeled ALL of the preprocessor blocks in these files

halotroop2288 committed:
Combine effectively duplicate blocks

halotroop2288 committed:
Overhaul README.md

halotroop2288 committed:
[CI] Don't ignore exports on tools-enabled workflows

halotroop2288 commented on drivers/unix/file_access_unix.cpp:
```diff
@@ -142,7 +143,7 @@ Error FileAccessUnix::_open(const String &p_path, int p_mode_flags) {
 #endif
 	}
-
+#endif // HORIZON_ENABLED
```
Typo: missing !

halotroop2288 commented on platform/switch/Info.plist:
```diff
@@ -0,0 +1,175 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Is this a generated file?

halotroop2288 commented on platform/switch/joypad_switch.cpp:
```diff
@@ -0,0 +1,96 @@
+	input = in;
+
+	// TODO: n players?
+	padConfigureInput(1, HidNpadStyleSet_NpadStandard);
```
Shouldn't this `1` be `JOYPADS_MAX`? After all, libNX is asking for `max_players`.
> max_players The maximum supported number of players (1 to 8).

Not doing this limits you to one player/controller. You could implement a way to ask the game developer how many are supported, but it's easier just to put 8 for now.

Better to allow more players than necessary than fewer.

halotroop2288 committed:
Add Switch notices to the changelog

halotroop2288 committed:
Added Vita notice to the changelog

halotroop2288 committed:
[Switch] Support up to eight controllers

halotroop2288 commented:
I have spent uhm... TWO weeks straight on this project, and I uh... ***completely*** lost track of time.
I'm gonna merge this within 24 hours if no further changes are requested.

halotroop2288 committed:
Delete old Switch scripts

A way to create these releases will return in the future!
